File: sdks/swift/Sources/Seda/Seda.swift
import Foundation

private let supportedProtocol = 1

public struct SedaStatus: Codable, Equatable, Sendable {
    public let name: String
    public let version: String
    public let `protocol`: Int
    public let ready: Bool
}

public struct SedaModelIdentity: Codable, Equatable, Sendable {
    public let id: String
    public let revision: String
    public let variant: String
    public let runtime: String
}

public struct SedaLanguageCapabilities: Codable, Equatable, Sendable {
    public let mode: String
    public let supported: [String]
    public let supportsAuto: Bool
    public let fixed: String?
}

public struct SedaCapabilities: Codable, Equatable, Sendable {
    public let runtime: String
    public let resolvedModel: SedaModelIdentity
    public let language: SedaLanguageCapabilities
    public let streaming: String
    public let punctuation: Bool
    public let wordTimestamps: Bool
    public let globalPushToTalk: Bool
    public let focusedAppInsertion: Bool
}

public struct SedaWord: Codable, Equatable, Sendable {
    public let text: String
    public let startMs: UInt64
    public let endMs: UInt64
    public let confidence: Float?
}

public struct SedaTranscript: Codable, Equatable, Sendable {
    public let text: String
    public let words: [SedaWord]
    public let language: String?
    public let durationMs: UInt64
}

public struct SedaTranscriptUpdate: Codable, Equatable, Sendable {
    public let segmentID: String
    public let revision: UInt64
    public let text: String
    public let stableText: String
    public let unstableText: String
    public let final: Bool
    public let words: [SedaWord]

    private enum CodingKeys: String, CodingKey {
        case segmentID = "segment_id"
        case revision
        case text
        case stableText = "stable_text"
        case unstableText = "unstable_text"
        case final
        case words
    }
}

public struct SedaError: Error, LocalizedError, Equatable, Sendable {
    public let code: String
    public let message: String
    public let recoverable: Bool
    public let status: Int?

    public var errorDescription: String? { message }
}

public final class Seda: @unchecked Sendable {
    private let baseURL: URL
    private let token: String
    private let urlSession: URLSession
    private let socketFactory: @Sendable (URL) -> any SedaWebSocket
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private init(
        baseURL: URL,
        token: String,
        urlSession: URLSession,
        socketFactory: @escaping @Sendable (URL) -> any SedaWebSocket
    ) {
        self.baseURL = baseURL
        self.token = token
        self.urlSession = urlSession
        self.socketFactory = socketFactory
    }

    public static func connect(
        baseURL: URL,
        token: String,
        urlSession: URLSession = .shared
    ) async throws -> Seda {
        guard !token.isEmpty else {
            throw SedaError(
                code: "invalid_request",
                message: "A bearer token is required",
                recoverable: false,
                status: nil
            )
        }
        let client = Seda(
            baseURL: baseURL,
            token: token,
            urlSession: urlSession,
            socketFactory: { url in
                URLSessionSedaWebSocket(session: urlSession, url: url)
            }
        )
        let status = try await client.status()
        guard status.protocol == supportedProtocol else {
            throw SedaError(
                code: "invalid_request",
                message: "Unsupported Seda protocol \(status.protocol); "
                    + "this client supports \(supportedProtocol)",
                recoverable: false,
                status: nil
            )
        }
        return client
    }

    static func testing(
        baseURL: URL,
        token: String,
        urlSession: URLSession,
        socketFactory: @escaping @Sendable (URL) -> any SedaWebSocket
    ) -> Seda {
        Seda(
            baseURL: baseURL,
            token: token,
            urlSession: urlSession,
            socketFactory: socketFactory
        )
    }

    public func status() async throws -> SedaStatus {
        try await request(path: "v1/status")
    }

    public func capabilities() async throws -> SedaCapabilities {
        try await request(path: "v1/capabilities")
    }

    public func transcribe(
        wav: Data,
        language: String? = nil
    ) async throws -> SedaTranscript {
        var components = URLComponents()
        components.path = "v1/transcriptions"
        if let language {
            components.queryItems = [URLQueryItem(name: "language", value: language)]
        }
        return try await request(
            path: components.string ?? "v1/transcriptions",
            method: "POST",
            body: wav,
            contentType: "audio/wav"
        )
    }

    public func listen(language: String? = nil) async throws -> SedaSession {
        var body: [String: Any] = [
            "input": [
                "encoding": "pcm_s16le",
                "sampleRate": 16_000,
                "channels": 1,
            ],
        ]
        if let language {
            body["language"] = language
        }
        let data = try JSONSerialization.data(withJSONObject: body)
        let created: SessionCreated = try await request(
            path: "v1/sessions",
            method: "POST",
            body: data,
            contentType: "application/json"
        )
        guard
            var components = URLComponents(
                url: created.websocketPath.hasPrefix("/")
                    ? URL(string: created.websocketPath, relativeTo: baseURL)!
                    : baseURL.appendingPathComponent(created.websocketPath),
                resolvingAgainstBaseURL: true
            )
        else {
            throw protocolError("Seda returned an invalid WebSocket path")
        }
        components.scheme = components.scheme == "https" ? "wss" : "ws"
        components.queryItems = [URLQueryItem(name: "ticket", value: created.ticket)]
        guard let socketURL = components.url else {
            throw protocolError("Seda returned an invalid WebSocket URL")
        }
        let socket = socketFactory(socketURL)
        socket.resume()
        return SedaSession(id: created.id, socket: socket)
    }

    private func request<T: Decodable>(
        path: String,
        method: String = "GET",
        body: Data? = nil,
        contentType: String? = nil
    ) async throws -> T {
        guard let target = URL(string: path, relativeTo: baseURL) else {
            throw protocolError("Invalid Seda request URL")
        }
        var request = URLRequest(url: target)
        request.httpMethod = method
        request.httpBody = body
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let contentType {
            request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        }
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await urlSession.data(for: request)
        } catch {
            throw SedaError(
                code: "runtime_failed",
                message: "Could not reach the Seda service: \(error.localizedDescription)",
                recoverable: true,
                status: nil
            )
        }
        guard let http = response as? HTTPURLResponse else {
            throw protocolError("Seda returned a non-HTTP response")
        }
        guard (200 ..< 300).contains(http.statusCode) else {
            throw decodeError(data, status: http.statusCode)
        }
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw protocolError("Seda returned invalid JSON")
        }
    }

    private func decodeError(_ data: Data, status: Int) -> SedaError {
        if
            let response = try? decoder.decode(ErrorEnvelope.self, from: data)
        {
            return SedaError(
                code: response.error.code,
                message: response.error.message,
                recoverable: response.error.recoverable,
                status: status
            )
        }
        return SedaError(
            code: "runtime_failed",
            message: "Seda request failed with HTTP \(status)",
            recoverable: status >= 500,
            status: status
        )
    }
}

public actor SedaSession {
    public nonisolated let id: String
    private let socket: any SedaWebSocket
    private let decoder = JSONDecoder()
    private var settled = false

    init(id: String, socket: any SedaWebSocket) {
        self.id = id
        self.socket = socket
    }

    public func write(_ pcmS16LE: Data) async throws {
        guard !settled else {
            throw protocolError("Session is already closed")
        }
        try await socket.send(.data(pcmS16LE))
    }

    public func commit(
        onTranscript: (@Sendable (SedaTranscriptUpdate) -> Void)? = nil
    ) async throws -> SedaTranscript {
        guard !settled else {
            throw protocolError("Session is already closed")
        }
        try await socket.send(.string(#"{"type":"commit"}"#))
        defer { close() }
        while true {
            let message = try await socket.receive()
            guard case let .string(text) = message, let data = text.data(using: .utf8)
            else {
                continue
            }
            let event: EventEnvelope
            do {
                event = try decoder.decode(EventEnvelope.self, from: data)
            } catch {
                throw protocolError("Seda returned an invalid live event")
            }
            switch event.type {
            case "transcript":
                if let onTranscript {
                    onTranscript(try decoder.decode(SedaTranscriptUpdate.self, from: data))
                }
            case "completed":
                guard let transcript = event.transcript else {
                    throw protocolError("Completed event has no transcript")
                }
                return transcript
            case "error":
                guard let error = event.error else {
                    throw protocolError("Malformed error event")
                }
                throw SedaError(
                    code: error.code,
                    message: error.message,
                    recoverable: error.recoverable,
                    status: nil
                )
            default:
                continue
            }
        }
    }

    public func cancel() async throws {
        guard !settled else { return }
        defer { close() }
        try await socket.send(.string(#"{"type":"cancel"}"#))
    }

    private func close() {
        settled = true
        socket.cancel()
    }
}

protocol SedaWebSocket: Sendable {
    func resume()
    func send(_ message: URLSessionWebSocketTask.Message) async throws
    func receive() async throws -> URLSessionWebSocketTask.Message
    func cancel()
}

private final class URLSessionSedaWebSocket: SedaWebSocket, @unchecked Sendable {
    private let task: URLSessionWebSocketTask

    init(session: URLSession, url: URL) {
        task = session.webSocketTask(with: url)
    }

    func resume() {
        task.resume()
    }

    func send(_ message: URLSessionWebSocketTask.Message) async throws {
        try await task.send(message)
    }

    func receive() async throws -> URLSessionWebSocketTask.Message {
        try await task.receive()
    }

    func cancel() {
        task.cancel(with: .normalClosure, reason: nil)
    }
}

private struct SessionCreated: Decodable {
    let id: String
    let websocketPath: String
    let ticket: String
}

private struct ErrorBody: Decodable {
    let code: String
    let message: String
    let recoverable: Bool
}

private struct ErrorEnvelope: Decodable {
    let error: ErrorBody
}

private struct EventEnvelope: Decodable {
    let type: String
    let transcript: SedaTranscript?
    let error: ErrorBody?
}

private func protocolError(_ message: String) -> SedaError {
    SedaError(
        code: "runtime_failed",
        message: message,
        recoverable: false,
        status: nil
    )
}

File: sdks/go/seda.go
// Package seda provides a typed client for Seda's local HTTP and WebSocket API.
package seda

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

const protocolVersion = 1

// Status describes a running Seda service.
type Status struct {
	Name     string `json:"name"`
	Version  string `json:"version"`
	Protocol int    `json:"protocol"`
	Ready    bool   `json:"ready"`
}

// ModelIdentity is the exact loaded model, revision, variant, and runtime.
type ModelIdentity struct {
	ID       string `json:"id"`
	Revision string `json:"revision"`
	Variant  string `json:"variant"`
	Runtime  string `json:"runtime"`
}

// LanguageCapabilities describes when and how a loaded model accepts language.
type LanguageCapabilities struct {
	Mode         string   `json:"mode"`
	Supported    []string `json:"supported"`
	SupportsAuto bool     `json:"supportsAuto"`
	Fixed        string   `json:"fixed,omitempty"`
}

// Capabilities describes the active model and service features.
type Capabilities struct {
	Runtime             string               `json:"runtime"`
	ResolvedModel       ModelIdentity        `json:"resolvedModel"`
	Language            LanguageCapabilities `json:"language"`
	Streaming           string               `json:"streaming"`
	Punctuation         bool                 `json:"punctuation"`
	WordTimestamps      bool                 `json:"wordTimestamps"`
	GlobalPushToTalk    bool                 `json:"globalPushToTalk"`
	FocusedAppInsertion bool                 `json:"focusedAppInsertion"`
}

// Word is one timestamped recognized word.
type Word struct {
	Text       string   `json:"text"`
	StartMS    uint64   `json:"startMs"`
	EndMS      uint64   `json:"endMs"`
	Confidence *float32 `json:"confidence,omitempty"`
}

// Transcript is a completed transcription.
type Transcript struct {
	Text       string `json:"text"`
	Words      []Word `json:"words"`
	Language   string `json:"language,omitempty"`
	DurationMS uint64 `json:"durationMs"`
}

// TranscriptUpdate is a revisable or final live result.
type TranscriptUpdate struct {
	SegmentID    string `json:"segment_id"`
	Revision     uint64 `json:"revision"`
	Text         string `json:"text"`
	StableText   string `json:"stable_text"`
	UnstableText string `json:"unstable_text"`
	Final        bool   `json:"final"`
	Words        []Word `json:"words"`
}

// Error is a stable Seda protocol failure.
type Error struct {
	Code        string
	Message     string
	Recoverable bool
	StatusCode  int
}

func (e *Error) Error() string { return e.Message }

type socket interface {
	WriteMessage(messageType int, data []byte) error
	ReadMessage() (messageType int, data []byte, err error)
	SetReadDeadline(time.Time) error
	Close() error
}

type socketDialer func(context.Context, string) (socket, error)

// Options configures a client connection.
type Options struct {
	BaseURL    string
	Token      string
	HTTPClient *http.Client
	dial       socketDialer
}

// Client is safe to reuse across complete and live transcriptions.
type Client struct {
	baseURL    *url.URL
	token      string
	httpClient *http.Client
	dial       socketDialer
}

// Connect verifies the service and returns a reusable client.
func Connect(ctx context.Context, options Options) (*Client, error) {
	if options.Token == "" {
		return nil, errors.New("seda: token is required")
	}
	baseURL, err := url.Parse(options.BaseURL)
	if err != nil || (baseURL.Scheme != "http" && baseURL.Scheme != "https") {
		return nil, errors.New("seda: BaseURL must use http or https")
	}
	if !strings.HasSuffix(baseURL.Path, "/") {
		baseURL.Path += "/"
	}
	client := &Client{
		baseURL:    baseURL,
		token:      options.Token,
		httpClient: options.HTTPClient,
		dial:       options.dial,
	}
	if client.httpClient == nil {
		client.httpClient = &http.Client{Timeout: 30 * time.Second}
	}
	if client.dial == nil {
		client.dial = func(ctx context.Context, target string) (socket, error) {
			connection, _, err := websocket.DefaultDialer.DialContext(ctx, target, nil)
			return connection, err
		}
	}
	status, err := client.Status(ctx)
	if err != nil {
		return nil, err
	}
	if status.Protocol != protocolVersion {
		return nil, fmt.Errorf(
			"seda: unsupported protocol %d; this client supports %d",
			status.Protocol,
			protocolVersion,
		)
	}
	return client, nil
}

// Status returns service health and protocol compatibility.
func (c *Client) Status(ctx context.Context) (Status, error) {
	var result Status
	err := c.request(ctx, http.MethodGet, "v1/status", nil, "", &result)
	return result, err
}

// Capabilities returns the exact resolved model and language behavior.
func (c *Client) Capabilities(ctx context.Context) (Capabilities, error) {
	var result Capabilities
	err := c.request(ctx, http.MethodGet, "v1/capabilities", nil, "", &result)
	return result, err
}

// TranscribeOptions are scoped to one complete audio request.
type TranscribeOptions struct {
	Language string
}

// Transcribe recognizes one mono PCM WAV file.
func (c *Client) Transcribe(
	ctx context.Context,
	wav []byte,
	options TranscribeOptions,
) (Transcript, error) {
	path := "v1/transcriptions"
	if options.Language != "" {
		path += "?language=" + url.QueryEscape(options.Language)
	}
	var result Transcript
	err := c.request(ctx, http.MethodPost, path, wav, "audio/wav", &result)
	return result, err
}

// ListenOptions are fixed when a new live stream begins.
type ListenOptions struct {
	Language string
}

type sessionCreated struct {
	ID            string `json:"id"`
	WebSocketPath string `json:"websocketPath"`
	Ticket        string `json:"ticket"`
}

// Listen opens a new language-scoped live stream against the prepared model.
func (c *Client) Listen(
	ctx context.Context,
	options ListenOptions,
) (*Session, error) {
	request := map[string]any{
		"input": map[string]any{
			"encoding":   "pcm_s16le",
			"sampleRate": 16_000,
			"channels":   1,
		},
	}
	if options.Language != "" {
		request["language"] = options.Language
	}
	body, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	var created sessionCreated
	if err := c.request(
		ctx,
		http.MethodPost,
		"v1/sessions",
		body,
		"application/json",
		&created,
	); err != nil {
		return nil, err
	}
	target := c.baseURL.ResolveReference(&url.URL{Path: created.WebSocketPath})
	if target.Scheme == "https" {
		target.Scheme = "wss"
	} else {
		target.Scheme = "ws"
	}
	query := target.Query()
	query.Set("ticket", created.Ticket)
	target.RawQuery = query.Encode()
	connection, err := c.dial(ctx, target.String())
	if err != nil {
		return nil, fmt.Errorf("seda: open live session: %w", err)
	}
	return &Session{id: created.ID, socket: connection}, nil
}

func (c *Client) request(
	ctx context.Context,
	method string,
	path string,
	body []byte,
	contentType string,
	output any,
) error {
	target, err := c.baseURL.Parse(path)
	if err != nil {
		return err
	}
	request, err := http.NewRequestWithContext(
		ctx,
		method,
		target.String(),
		bytes.NewReader(body),
	)
	if err != nil {
		return err
	}
	request.Header.Set("Authorization", "Bearer "+c.token)
	if contentType != "" {
		request.Header.Set("Content-Type", contentType)
	}
	response, err := c.httpClient.Do(request)
	if err != nil {
		return fmt.Errorf("seda: request failed: %w", err)
	}
	defer response.Body.Close()
	payload, err := io.ReadAll(io.LimitReader(response.Body, 16<<20))
	if err != nil {
		return err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return responseError(payload, response.StatusCode)
	}
	if err := json.Unmarshal(payload, output); err != nil {
		return fmt.Errorf("seda: invalid JSON response: %w", err)
	}
	return nil
}

// Session is one independent live transcription stream.
type Session struct {
	id      string
	socket  socket
	settled bool
}

// ID returns the server-generated session identifier.
func (s *Session) ID() string { return s.id }

// Write sends 16 kHz mono signed little-endian PCM.
func (s *Session) Write(pcmS16LE []byte) error {
	if s.settled {
		return errors.New("seda: session is already closed")
	}
	return s.socket.WriteMessage(websocket.BinaryMessage, pcmS16LE)
}

// Commit finalizes the stream. onTranscript receives every revisable result.
func (s *Session) Commit(
	ctx context.Context,
	onTranscript func(TranscriptUpdate),
) (Transcript, error) {
	if s.settled {
		return Transcript{}, errors.New("seda: session is already closed")
	}
	if err := s.socket.WriteMessage(
		websocket.TextMessage,
		[]byte(`{"type":"commit"}`),
	); err != nil {
		return Transcript{}, err
	}
	defer s.close()
	if deadline, ok := ctx.Deadline(); ok {
		_ = s.socket.SetReadDeadline(deadline)
	}
	for {
		messageType, payload, err := s.socket.ReadMessage()
		if err != nil {
			return Transcript{}, fmt.Errorf("seda: read live event: %w", err)
		}
		if messageType != websocket.TextMessage {
			continue
		}
		var envelope struct {
			Type       string         `json:"type"`
			Transcript *Transcript    `json:"transcript"`
			Error      *protocolError `json:"error"`
		}
		if err := json.Unmarshal(payload, &envelope); err != nil {
			return Transcript{}, fmt.Errorf("seda: invalid live event: %w", err)
		}
		switch envelope.Type {
		case "transcript":
			if onTranscript != nil {
				var update TranscriptUpdate
				if err := json.Unmarshal(payload, &update); err != nil {
					return Transcript{}, err
				}
				onTranscript(update)
			}
		case "completed":
			if envelope.Transcript == nil {
				return Transcript{}, errors.New("seda: completed event has no transcript")
			}
			return *envelope.Transcript, nil
		case "error":
			if envelope.Error == nil {
				return Transcript{}, errors.New("seda: malformed error event")
			}
			return Transcript{}, envelope.Error.toError(0)
		}
	}
}

// Cancel closes a stream without producing a final transcript.
func (s *Session) Cancel() error {
	if s.settled {
		return nil
	}
	defer s.close()
	return s.socket.WriteMessage(
		websocket.TextMessage,
		[]byte(`{"type":"cancel"}`),
	)
}

func (s *Session) close() {
	s.settled = true
	_ = s.socket.Close()
}

type protocolError struct {
	Code        string `json:"code"`
	Message     string `json:"message"`
	Recoverable bool   `json:"recoverable"`
}

func (e protocolError) toError(status int) *Error {
	return &Error{
		Code:        e.Code,
		Message:     e.Message,
		Recoverable: e.Recoverable,
		StatusCode:  status,
	}
}

func responseError(payload []byte, status int) error {
	var envelope struct {
		Error protocolError `json:"error"`
	}
	if err := json.Unmarshal(payload, &envelope); err == nil &&
		envelope.Error.Message != "" {
		return envelope.Error.toError(status)
	}
	return &Error{
		Code:        "runtime_failed",
		Message:     fmt.Sprintf("Seda request failed with HTTP %d", status),
		Recoverable: status >= 500,
		StatusCode:  status,
	}
}
